jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/StageActivationConditionTest.java

```diff
@@ -92,8 +92,8 @@ public void testAutoCompleteCheckActivationData() {
                                 @Override
                                 public Boolean execute(Context context) {
                                     KieSession ksession = ((RegistryContext) context).lookup( KieSession.class );
-                                    
-                                    WorkflowProcessInstance pi = (WorkflowProcessInstance) ksession.getProcessInstance(processInstanceId, true);
+
```

Comment:
Why is this change needed?

Reply:
getVariable of node instance requires read/write process. if you get this read only the test will fail.

Reply:
why getVariable requires read/write process?

Reply:
@mswiderski 
I cannot tell you why some operations were designed to work with read only or read / write. That is not documented anywhere.
The reason is when you disconnect the process this condition is fullyfilled
https://github.com/kiegroup/jbpm/blob/7.20.0.Final/jbpm-flow/src/main/java/org/jbpm/process/instance/impl/ProcessInstanceImpl.java#L97

causing failure in here
https://github.com/kiegroup/jbpm/pull/1471/files/e13dc20a8a8be0513b9aab4a1553d52703b31b3f#diff-6c5262fecaf86cc99369126786aad13cR97

Reply:
what I am asking is why this is needed? It wasn't needed before and I don't see a point in writing anything to process instance info when just reading it's variables. So could you point me what is actually making to fail?

Reply:
the problem in the jira is that the listener were registed for a read only process instance. This was because while in some parts of the code the process instance reconnected then after being used was disconnected again.

That was happening here
https://github.com/kiegroup/jbpm/blob/7.19.0.Final/jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java#L174

The collateral effect is that some operations that were being relying on the process instance were connected now are failing. Disconnecting the process instance causes the state variables to be null therefore you cannot navigate to the getProcess (this happens internally during getVariable)

stack trace for the getting variable
~~~
at org.jbpm.process.instance.impl.ProcessInstanceImpl.getProcess(ProcessInstanceImpl.java:97)
	at org.jbpm.workflow.instance.impl.WorkflowProcessInstanceImpl.getWorkflowProcess(WorkflowProcessInstanceImpl.java:292)
	at org.jbpm.workflow.instance.impl.WorkflowProcessInstanceImpl.getNodeContainer(WorkflowProcessInstanceImpl.java:124)
	at org.jbpm.workflow.instance.impl.NodeInstanceImpl.getNode(NodeInstanceImpl.java:134)
	at org.jbpm.workflow.instance.impl.NodeInstanceImpl.resolveContext(NodeInstanceImpl.java:434)
	at org.jbpm.workflow.instance.impl.NodeInstanceImpl.resolveContextInstance(NodeInstanceImpl.java:441)
	at org.jbpm.workflow.instance.impl.NodeInstanceImpl.getVariable(NodeInstanceImpl.java:487)
~~~

Reply:
did you track down what is calling disconnect during the execution? It should only be invoked upon completion ... usually transaction completion.

Reply:
@mswiderski giving your comment the this makes me think that probably this fix is incorrect. I was following the logic based on this

https://github.com/kiegroup/jbpm/blob/7.20.0.Final/jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/ProcessInstanceInfo.java#L194

and copying in the in to the JPAProcessInstanceManager causing the disconnection. The disconnection causes removing the listener at several stages including the one giving problem.

Reply:
@elguardian In general what I don't see a need is for reading operations we should not make modifications to the underlying entity - process instance info. hat's why I want to understand why we need to do it now. The main reason for it is that it will increase probability for optimistic lock exceptions as even getting process variables will update the version.
With that said, mixing readonly and write operations on process instance should be work as expected - as if I recall correctly readonly does the disconnect as you pointed out but also removes from the internal map of loaded instances and thus should be reloaded when write operation wants to access it. Maybe that would be the area to investigate more?